services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedDeviceDao.java

```diff
@@ -372,23 +373,39 @@ public void registerLivenessChecks(final HealthCheckHandler livenessHandler) {
                 .start();
         resourceVersion.ifPresent(v -> TracingHelper.TAG_RESOURCE_VERSION.set(span, v));
 
-        final JsonObject updateDeviceQuery = MongoDbDocumentBuilder.builder()
-                .withVersion(resourceVersion)
-                .withTenantId(deviceConfig.getTenantId())
-                .withDeviceId(deviceConfig.getDeviceId())
-                .document();
-
-        final Promise<JsonObject> updateDevicePromise = Promise.promise();
-
-        mongoClient.findOneAndUpdateWithOptions(
-                collectionName,
-                updateDeviceQuery,
-                MongoDbDocumentBuilder.builder().forUpdateOf(deviceConfig).document(),
-                new FindOptions(),
-                new UpdateOptions().setReturningNewDocument(true),
-                updateDevicePromise);
+        return getById(deviceConfig.getTenantId(), deviceConfig.getDeviceId())
+                .compose(currentDeviceConfig -> {
+                    final MongoDbBasedDeviceDto dto = MongoDbBasedDeviceDto.forUpdate(
+                            // use creation date from DB as this will never change
+                            () -> currentDeviceConfig,
+                            // but copy all other (updated) data from DTO that has been passed in
+                            deviceConfig.getData(),
+                            deviceConfig.getVersion());
+                    // we need to manually copy the device status because the passed in DeviceDto's data
+                    // property does not contain any status information (it had been removed by the
+                    // DeviceDto.fromUpdate method used to create it)
```

Comment:
> You mean DeviceDto.forUpdate here?

yes, may bad

> The dto.setDeviceStatus(deviceConfig.getDeviceStatus()); line is about adopting the deviceConfig autoProvisioned and notificationSent values, not the creationTime or lastUpdate (because they are always null here), right?

yes

> It's a bit confusing, that the dto object has 2 sets of creationTime and lastUpdate values in it: once in the dto object fields and once in its deviceStatus object.

Tell me about it ;-) Let me see what I can do ...

Reply:
I have pushed changes but I'd rather address the issue with the duplicate DeviceStatus in a future PR.